### PR TITLE
WebGLRenderer: Add support for uv3 and uv4.

### PR DESCRIPTION
**Pull request by mrdoob**
Related issue: #25721

**Description**

Support up to 4 uv channels instead of just 2.
Also moved attribute definitions to `WebGLProgram.js`.

### CODE REVIEW

File: src/renderers/webgl/WebGLPrograms.js [hunk 1]
```diff
@@ -38,6 +38,8 @@ function WebGLPrograms( renderer, cubemaps, cubeuvmaps, extensions, capabilities
 	function getChannel( value ) {
 
 		if ( value === 1 ) return 'uv2';
```
Comment (LeviPesin) [inline src/renderers/webgl/WebGLPrograms.js:40]: Maybe just `if ( value > 0 ) return 'uv' + ( value + 1 );`?

Comment (mrdoob) [reply src/renderers/webgl/WebGLPrograms.js:40]: The current code is more readable.

**Review by Mugen87 — APPROVED** [review]

**Comment by mrdoob** [comment]
I really dislike the `uv`, `uv2`, `uv3`, `uv4` names...

But I don't know how many things, if any, would be to rename them now.

And if we did, should it be `uv_0`? `texcoord_0`? ...?

**Comment by Mugen87** [comment]
> But I don't know how many things, if any, would be to rename them now.

I'm afraid too many things depend on these names. But if we could start from scratch I would favor the `texcoord_0` semantic.

**Comment by pailhead** [comment]
uv_0 sounds good.

**Comment by donmccurdy** [comment]
I slightly prefer "uv" over "texcoord". I think the glTF term is less common in this case, and we use the word UV elsewhere in our API, not just the attribute name. It would be great for the index used in the name to match up with the value assigned to `texture.channel`, and not be off by 1.

Do we want to mix underscores with camel case though? If we someday add an additional set of skinning weights, do we prefer `skinWeight2` or `skinWeight_2` or ...?

**Comment by WestLangley** [comment]
>I really dislike the uv, uv2, uv3, uv4 names...

Me, too.

How about

```js
geometry.attributes.uvs[ i ]
```
with `.uv` and `.uv2` retained as getters for backwards compatibility.

**Comment by pailhead** [comment]
@WestLangley  I think it can’t just be a simple array since then you could make it sparse? Some api (add, remove) could be more robust.

**Comment by gkjohnson** [comment]
I personally do not see the issue with `uv`, `uv2`, etc names or why an underscore or changing to "texcoord" would be an improvement. Either way all suggestions and current implementation all seem understandable to me.

Whatever the case, though, please do not use an array of BufferAttributes in the attributes dictionary:

> ```js
> geometry.attributes.uvs[ i ]
> ```
> with .uv and .uv2 retained as getters for backwards compatibility.

I've often generically iterated over the attribute fields to process them into textures, merge geometries, compress data, etc. Adding a new object structure in `attributes` doesn't add a lot of value and will needlessly break and complicate user code. There's also a level of simplicity with attribute dictionary names directly mapping to attribute value names in shader code.

Please also consider that changing the name of `uv` will break user land shaders that use that field in vertex programs.

**Comment by pailhead** [comment]
What about a mechanism on the geometry not the attributes? Not sure how it would look like, but this doesn’t feel like this belongs on the attributes, rather above it, managing them.

**Comment by Mugen87** [comment]
I fully agree with what @gkjohnson said 👍 .

**Comment by mrdoob** [comment]
I did a bad job at explaining what it is that I don't like 😅

When I implemented `texture.channel` I made it so it follows GLTF's TEXCOORD ids but the Three.js developer it'll be confusing because `texture.channel` doesn't map the ids used in the attribute.

| GLTF | channel | attribute |
| --- | --- | --- |
| TEXCOORD_0 | 0 |  uv  |
| TEXCOORD_1 | 1 |  uv2 | 
| TEXCOORD_2 | 2 |  uv3 |
| TEXCOORD_3 | 3 |  uv4  |

**Option 1**
One solution would be to make `texture.channel` start by 1 instead.
But that could be confusing when authoring content.

| GLTF | channel | attribute |
| --- | --- | --- |
| TEXCOORD_0 | 1 |  uv  |
| TEXCOORD_1 | 2 |  uv2 | 
| TEXCOORD_2 | 3 |  uv3 |
| TEXCOORD_3 | 4 |  uv4  |

**Option 2**
Another solution is to rename the attributes (making sure we don't break backwards compatibility)

| GLTF | channel | attribute |
| --- | --- | --- |
| TEXCOORD_0 | 0 |  uv0  |
| TEXCOORD_1 | 1 |  uv1 | 
| TEXCOORD_2 | 2 |  uv2 |
| TEXCOORD_3 | 3 |  uv3  |

I think I prefer the second option and if people is okay with it I can give it a try after merging this PR.

**Comment by donmccurdy** [comment]
Agreed. I like both of those, and also have a small preference for option 2. No need to change the underlying geometry/attribute API.

**Comment by pailhead** [comment]
See if you can give it some more thought. I don’t think that this belongs in the attributes, attributes are just attributes. And while texcoord seems to be a GL term, it soon changes to UV. Kinda curious about that historically.

But, since it is doing some magic - some in a sense - more than the magic when you name your buffer ‘foo’ I feel like it could be explored a bit how it would look like if geometry managed this.

more maybe have some builder helper. There is no reason why someone would dynamically alter - remove or add uv attributes at runtime.

**Comment by makc** [comment]
why not just write an attribute name (string) into texture.channel @mrdoob

**Comment by pailhead** [comment]
Hmmm, you would have to “turn it off” for its actual name? So that you don’t get “horse” and uv0? But I kinda like this idea.

**Comment by pailhead** [comment]
Hmmm, you would have to “turn it off” for its actual name? So that you don’t get “horse” and uv0? But I kinda like this idea.

**Comment by makc** [comment]
@pailhead like, if users want to use the attribute named "horse" - let them. why not? what's the big deal? just set it to "uv" by default, and to "uvX" with correct X for the maps that require another layer.

**Comment by pailhead** [comment]
I see, if you write “horse” to an array, you would still call setAttribute(“horse”…) but you would actually get uv_x and no horse?

**Comment by makc** [comment]
why array? I was talking about apparently numeric property of Texture that mrdoob modelled after gltf spec. apparently.

**Comment by makc** [comment]
e.g. if the texture is marked to use "horse" (or "uv3"), three would check if there is such an attribute (and it normally will be, because the loader loaded uv3, or the user added the horse) and if not - throw some error at which point the user knows his code needs fixing.

**Comment by pailhead** [comment]
I misunderstood something probably.

**Comment by makc** [comment]
yep.

**Comment by pailhead** [comment]
As an interface that makes sense to me. Feels like it would be a massive change.

**Comment by pailhead** [comment]
Actually, i don't think it would be such a massive change. Such an list would still mean that textures are used via
```
attribute vec3 horse;
varying vec2 vUv0;

... 

  vUv0 = horse.xy;
 ```
 
 Could actually even eliminate a lot of this duplicated code that deals with injecting `attribute vec2 uv_x`.

**Comment by makc** [comment]
more massive than renaming uv to uv_0 for no reason? 😅 like, people will need to edit their shaders and stuff

**Comment by makc** [comment]
>   vUv0 = horse.xy;

just do 'v' + capitalize(attributeName). so that uv turns to vUv (like now? iirc), uv3 to vUv3, horse to vHorse

**Comment by pailhead** [comment]
I was thinking at least the fragment shaders wouldn’t have to be changed. And there is no need to prefix anything just point an attribute to where uv would have been accessed

**Comment by makc** [comment]
but then you have to order them somehow. and you would also have a failure if users specify 5 channels.

**Review by donmccurdy — APPROVED** [review]

**Pull request merged**